# Chrom-Lasso
## Introduction of Chrom-Lasso
Chrom-Lasso is a tool to analyze Hi-C data for identifying cis chromatin interactions (interactions occur within chromosome).
## Introduction of Folders
The following folders are code, input files, test data, and tutorial for Chrom-Lasso.
1. Code：
This folder contains codes for Chrom-Lasso, the usage of them is shown in the "Tutorial" folder.
2. Prepare_Input_File：
This folder contains input files needed by Chrom-Lasso to do analysis, mainly are domain files and cutting site files for Mouse and Human,
The domain files are constant for specific species, but the genomic coordinate in the file can be changed with genome version by tool "LiftOver".
The cutting site files can be produced by tool "Oligomatch" with reference genome and cutting sequence for the restriction endonuclease.
3. Test_Data：
This folder contains "sortChr" files for Mouse and Human for testing Chrom-Lasso, the "sortChr" file are preprocessed file from raw "fastq" sequencing files,
the preprocessing of "fastq" files can be done with our tutorial, which is the same as "Juicer", so you can use "Juicer" as well.
The "Juicer" generates "merged_nodups" file, and you can use the Shell scripts in tutorial that sorts the file by chromosome order to generate "sortChr" file for further analysis.
4. Tutorial：
This folder contains the analysis pipeline for Mouse and Human, you can directly use "sotrChr" files in "Test_Data" folder to test the pipeline, 
you can also use your own "fastq" files from Hi-C experiments to test the tutorial step by step from the preprocessing to running Chrom-Lasso.
## Environment
The compile of Chrom-Lasso recommends: gcc (4.9.2), boost_1.51. And it also needs R (>=3.0) to run polynomial regression and lasso regression. Users can use "makefile" in Code folder for compile.
## Tutorial
### Prepare input files
Chrom-Lasso needs 2 input files prepared according to the Hi-C experimental design. 
#### Cutting site file
The cutting site file contains the cutting sites of restriction enzyme used in Hi-C experiments. Each line stands for a cutting site locus on the genome.<br>  
![cutting site file](https://github.com/Lan-lab/Chrom-Lasso/blob/main/documentation/cutting_site_bed.png)<br>  
This bedfile can be generated by an open source tool OligoMatch following the instruction in /Prepare_Input_File/Cutting_Site_File/Prepare_CuttingSite_File_By_Oligomatch.
#### Domain file
The domain file contains the genomic regions identified as TADs. Each line stands for a TAD identified.<br>  
![domain file](https://github.com/Lan-lab/Chrom-Lasso/blob/main/documentation/domain_file.png)<br>
The raw domain files are downloaded from: (Dixon, J., Selvaraj, S., Yue, F. et al. Topological domains in mammalian genomes identified by analysis of chromatin interactions. 
Nature 485, 376–380 (2012)). The raw version for Mouse genome is mm9 and for Human genome is hg18. Users can transform the genome version via tool "LiftOver". 
#### sortChr file
Users should first generate merged_nodups.txt file by JUICER (https://github.com/aidenlab/juicer) from raw Hi-C sequencing fastq files.<br>  
Here, we use Mouse_merged_nodups.txt file as example to generate sortChr file<br>
```
awk '{if($1==0){strand1=1;}else if($1==16) {strand1=0;} if($5==0){strand2=1;}else if($5==16) {strand2=0;} print $2"\t"$3"\t"strand1"\t"$6"\t"$7"\t"strand2"\t0";}' Mouse_merged_nodups.txt > Mouse.formatted
for chr in {1..19} X;do awk '{if($1=="'$chr'") print $0;}' Mouse.formatted;done > Mouse.sortChr
```
The sortChr file contains the paired end sequencing information of Hi-C data.<br>
![sortChr file](https://github.com/Lan-lab/Chrom-Lasso/blob/main/documentation/sortChr.png)<br>

 
